### PR TITLE
[tanjiro] Round-6: Lion+RFF + EMA decay 0.9999 (yi #13 compounding)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

EMA decay 0.9999 (vs current 0.9995) will reduce test_abupt further on the Lion+RFF config. A slower-decaying EMA accumulates a longer effective history of model weights, sharpening the ensemble's calibration and reducing prediction variance — particularly for the anisotropic tau fields (tau_y, tau_z) where variance is highest. This is a direct port of yi's PR #13 (norman) finding: progressive EMA 0.99→0.9999 gave −9% on every axis. We're applying the terminal value (0.9999) directly since our 270min run only fits ~9 uncompiled epochs.

**Single delta from edward #51 baseline (Lion+RFF+ema=0.9995):** only `--ema-decay 0.9999`.

**Why this has a chance:** EMA decay is optimizer-orthogonal — it acts on the weight trajectory after optimization, not on gradients. There is no Lion-fragility risk. If edward #51 reproduces val 10.665 (likely), then 0.9999 EMA acts on the same trajectory with a longer averaging horizon. The yi #13 finding was one of the largest single-lever improvements in the yi branch.

## Instructions

Run the following command in `target/`:

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro \
  --wandb-name "tanjiro/round6-lion-rff-ema9999-rank0" \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --rff-num-features 32 --rff-sigma 1.0 \
  --no-compile-model \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9999 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

**Key changes vs edward #51 baseline:**
- `--ema-decay 0.9999` (was `0.9995`)
- Everything else identical to edward #51 (`iocqp761` / `ftg0ci0p`)

**Do NOT use `--compile-model`.** Lion+compile diverges at any LR (9 confirmed failures). Keep uncompiled.

When the run completes, post:
1. W&B run URL
2. All `test_primary/*` values from the W&B summary
3. Comparison table vs arm B SOTA (11.303) and edward #51 interim val (10.665)

## Baseline

Current SOTA: **Lion lr=5e-5/wd=5e-4 uncompiled** (tanjiro arm B, W&B run `vnb7oheo`)

| Metric | Current SOTA | Edward #51 interim val | AB-UPT ref |
|---|---:|---:|---:|
| **abupt_mean** | **11.303** | **10.665 (val)** | — |
| surface_pressure | 6.216 | 6.738 | 3.82 |
| wall_shear | 11.315 | 12.087 | 7.29 |
| volume_pressure | 12.755 | 6.220 | 6.08 |
| tau_x | 9.563 | 10.301 | 5.35 |
| tau_y | 13.831 | 14.305 | 3.65 |
| tau_z | 14.147 | 15.762 | 3.63 |

**Target:** beat 11.303 test_abupt. Edward #51's val 10.665 suggests Lion+RFF already lands <11.30 test — adding EMA 0.9999 should squeeze further.

Reproduce command for arm B SOTA:
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --no-log-gradient-histograms
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

Picked up the run. The exact PR command is already running on this pod since 11:13:45 UTC.

- W&B rank-0 run: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ocubvc23
- Run name: `tanjiro/round6-lion-rff-ema9999-rank0-rank0`
- Status (~45 min in): Epoch 2/50 ~30%, val after epoch 1 = 85.06% (early-epoch noise; expected to drop fast over the next epochs).
- Per-epoch wall time: ~30 min uncompiled, matching edward #51 expectation. 270-min train cap → ~8-9 useful epochs before the budget kicks in, plus full_val + test rerun on best checkpoint.

Will report final `test_primary/*` once the run finishes.

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Results

**Hypothesis FAILED — clear negative result.** EMA decay 0.9999 (vs 0.9995 baseline) on Lion+RFF degrades every metric by ~2.5–3× because the slow EMA never finishes warming up within the 270-min / ~23K-step budget. Best-val checkpoint comes from a heavily biased EMA snapshot still dominated by random-init weights.

### Headline test_primary metrics (best-val EMA reload)

| Metric | This run (ema=0.9999) | Edward #51 (ema=0.9995) | SOTA Lion arm B | AB-UPT ref | Δ vs SOTA |
|---|---:|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **30.203** | 11.741 | **11.303** | — | **+167%** |
| `surface_pressure_rel_l2_pct` | 20.821 | 6.394 | 6.216 | 3.82 | +235% |
| `wall_shear_rel_l2_pct` | 31.519 | 11.837 | 11.315 | 7.29 | +179% |
| `volume_pressure_rel_l2_pct` | 25.057 | 13.111 | 12.755 | 6.08 | +96% |
| `wall_shear_x_rel_l2_pct` | 26.216 | — | 9.563 | 5.35 | +174% |
| `wall_shear_y_rel_l2_pct` | 39.381 | — | 13.831 | 3.65 | +185% |
| `wall_shear_z_rel_l2_pct` | 39.543 | — | 14.147 | 3.63 | +180% |

### Per-epoch val curve — this is the smoking gun

Same architecture, same optimizer, same RFF, same 9 epochs. Only EMA decay differs:

| Step | OURS (ema=0.9999) | EDWARD #51 (ema=0.9995) |
|---:|---:|---:|
| 2720 | 85.061 | 72.959 |
| 5440 | 80.977 | 35.148 |
| 8160 | 75.572 | 20.581 |
| 10880 | 63.353 | 15.788 |
| 13600 | 55.139 | 13.529 |
| 16320 | 47.076 | 12.312 |
| 19040 | 38.387 | 11.564 |
| 21760 | 32.053 | 11.008 |
| 23357 | **29.540** | **10.703** |

EMA at decay=0.9999 has effective horizon `1/(1-d) = 10000` steps. Best-checkpoint is selected from an EMA snapshot that has only seen ~2.3 effective horizons of history — still heavily biased toward the initial random weights. Edward #51's ema=0.9995 has horizon=2000 → fully warmed up by step ~6000. Train loss converged identically in both runs (`train/epoch_loss=0.0306` here vs ~0.030 for edward #51 — within noise), confirming the underlying model trained fine; only the EMA snapshot is broken.

### Why the yi #13 finding didn't transfer

The PR's hypothesis cited yi #13 (progressive 0.99 → 0.9999, −9% on every axis). The terminal value is not the lever — the **schedule** is. Yi started at 0.99 (horizon=100) so the EMA tracked the model closely from step 0, then ramped up to 0.9999 only after the EMA was already warmed up. Setting decay=0.9999 from step 50 with only 23K total steps is the bias-warmup pathology, not the smoothing benefit yi observed.

### Run details

- **W&B run:** https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ocubvc23 (rank-0)
- **Run name:** `tanjiro/round6-lion-rff-ema9999-rank0-rank0`
- **Best epoch:** 9 / 9 used (270-min cap)
- **Total train minutes:** 270.76
- **Global step at best:** 23357
- **Peak GPU memory:** ~74.5 GB / GPU (`system.gpu.*.memoryAllocatedBytes`), ~72.5% utilized — same envelope as edward #51 / Lion arm B baseline
- **Best-checkpoint source:** `ema` (selected from EMA, hence the bias)

### Exact command used

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro \
  --wandb-name "tanjiro/round6-lion-rff-ema9999-rank0" \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --rff-num-features 32 --rff-sigma 1.0 \
  --no-compile-model \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9999 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

### What happened

Hypothesis was that a slower EMA (longer averaging horizon) would sharpen calibration on the variance-heavy tau channels. The opposite happened on every metric. Root cause is straightforward EMA warmup bias: with effective horizon 10000 steps and only 23357 total training steps starting from `ema_start_step=50`, the EMA is dominated by the early random-init weights at every checkpoint we evaluate. The training itself was healthy (`train/epoch_loss` curve identical to edward #51 within noise), but `best_checkpoint/source=ema` means we're reading off a still-warming-up snapshot. The single-lever-change framing was not optimizer-orthogonal in the way the PR predicted — it interacts with the wall-clock budget through the EMA warmup transient.

### Suggested follow-ups

1. **Progressive cosine EMA 0.99 → 0.9999** (the actual yi #13 lever). Schedule keeps EMA closely tracking the model early, then smooths only once the model is near its plateau. Needs a small `train.py` patch to interpolate `ema_decay` over training; cleanly testable as a single-lever change vs Lion+RFF SOTA.
2. **Bias-corrected EMA** — divide by `1 - d^step` à la Adam. Cheap one-liner; would let `0.9999` work even with short budgets. Worth a quick A/B with the existing best-val selection logic.
3. **Eval raw-vs-EMA** — turn on `--eval-raw-vs-ema` so the best-checkpoint logic can pick the raw weights when EMA is still warming. Current code has the flag (`eval_raw_vs_ema: False`); flipping it would have caught this failure mode automatically and likely landed in the 11–12 range.
4. **Larger EMA start-delay** — `ema_start_step=50` is essentially "from the beginning". Pushing it to e.g. 5000 would let the model warm up before EMA begins accumulating; less powerful than (1) but a one-flag change.
5. **More epochs / compile fix for Lion** — orthogonal to EMA: if we could get 16+ epochs of Lion in budget, terminal-value-only EMA=0.9999 *might* finally start beating 0.9995. Currently blocked by the Lion+compile divergence (9 confirmed failures).

Decision-relevant takeaway for the advisor: the terminal EMA value alone is not the lever; the schedule is. Recommend pivoting to progressive EMA (option 1) and dropping flat 0.9999 from the queue.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #90 because ADVISOR: Confirmed clear dead end — closing.

**Result:** test_abupt 30.203 vs SOTA 11.208 = +169% regression. EMA decay 0.9999 fundamentally incompatible with 270m / 9-epoch budget — the slow EMA never warms up in the available step count, so best-val checkpoint reloads a heavily biased early-init weight average.

**Lesson:** EMA decay 0.9999 needs ~10× more steps (90+ epochs) before the EMA snapshot dominates the random-init contribution. Budget-incompatible regardless of base optimizer or RFF. EMA decay 0.9995 (current SOTA) is the right setting at this step count.

**Closed-door:** EMA in [0.9999, 0.99999] range is closed for this budget. Lower-decay variants (0.999, 0.99) may be worth exploring — they would let the EMA track the live weights more closely while still smoothing late-stage oscillations. Will revisit if other levers plateau.

Round 7 schedule sweep is the active focus (cosine T_max=24/50, warmup, RFF freq sweep). Reassigning to a new hypothesis..
